dwh/query_lib_big_query.py

```diff
@@ -0,0 +1,325 @@
+          L.location.LocationId AS locationId,
+          C.system AS encTypeSystem,
+          C.code AS encTypeCode,
+          L.location.display AS locationDisplay,
+          COUNT(*) AS num_encounters,
+          MIN(S.period.start) AS firstDate,
+          MAX(S.period.end) AS lastDate
+          from S, UNNEST(s.type.array) AS T, UNNEST(T.coding.array) AS C
+          LEFT JOIN UNNEST(s.location.array) AS L
+          {where}
+          GROUP BY S.subject.PatientId, L.location.LocationId,
+          L.location.display, C.system, C.code
+          {sample_count}
+    """
+
+        where_clause = self._construct_encounter_constraint(
+            self._enc_constraint
+        )
+        sql = sql_template.format(
+            table_name=table_name,
+            base_url=base_url,
+            data_set=bq_dataset,
+            where=where_clause,
+            sample_count=""
+            if sample_count is None
+            else "LIMIT " + str(sample_count),
+        )
+        return sql
+
+    def _build_obs_encounter_query(
+        self, dataset: str, base_url: str, sample_count: tp.Optional[int] = None
+    ) -> str:
+        """
+        Helper functions that builds the query to get patient_observation data
+        Args:
+            dataset: Indicates the schema to use.
+            sample_count: Restricts returned number of rows to sample count.
+                          Useful for testing and exploration.
+        """
+
+        sql_template = """
+    with O AS
+    (SELECT * FROM `{dataset}.Observation`),
+      O1 as
+      (SELECT
+          OC.system obs_system,
+          OC.code AS obs_code_coding_code,
+          OVC.code obs_value_code,
+          O.effective.dateTime obs_effective_datetime,
+          O.value.quantity.value AS obs_value_quantity,
+          O.subject.PatientId obs_subject_patient_id,
+          O.context.encounterId obs_context_encounter_id,
+          FORMAT('%s,%s', cast(O.effective.dateTime AS string),
+                 cast(O.value.quantity.value AS string)) AS date_and_value,
+          FORMAT('%s,%s', cast(O.effective.dateTime AS string), OVC.code)
+                 AS date_and_value_code,
+          FROM O LEFT JOIN UNNEST(O.code.coding.array) AS OC LEFT JOIN
+          UNNEST(O.value.codeableConcept.coding.array) AS OVC
+          where {code_coding_system} and {value_codeable_coding_system}
+          and {all_obs_constraints}
+          ),
+      E1 AS (
+          SELECT replace(E.id, '{base_url}', '') AS encounterId, C.system, C.code,
+            L.location.LocationId, L.location.display,
+            FROM `{dataset}.Encounter` AS E LEFT JOIN UNNEST(E.type.array) AS T LEFT JOIN
+            UNNEST(T.coding.array) AS C LEFT JOIN UNNEST(E.location.array) AS L
+            {encounter_where_clause}
+            ),
+      G AS (SELECT
+          obs_subject_patient_id patient_id,
+          obs_code_coding_code AS coding_code, -- TODO: Should this be coding.code
+          count(*) AS num_obs,
```

Comment:
```suggestion
          COUNT(*) AS num_obs,
```
also for all other SQL aggregate functions.